### PR TITLE
pr-upload: Upload bottles to the Internet Archive

### PR DESCRIPTION
**Pull request by sjackman**
`--archive-item=…` specifies the item identifier.
`HOMEBREW_INTERNET_ARCHIVE_KEY=access:secret` specifies the S3-like key. See https://archive.org/account/s3.php

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

### CODE REVIEW

**Comment by jonchang** [comment]
Multiple cloud storage services support an S3-like API, including Amazon itself. Would it perhaps make more sense to have a `utils/s3` that implements the needed APIs, then (if necessary) have `utils/s3/archive` to subclass/override the bits that are specific to e.g. the internet archive?

**Comment by sjackman** [comment]
This implementation could pretty easily be adapted to a true S3 implementation like Amazon S3, but in its current form will not work out-of-the-box. I personally have no immediate use case for uploading bottles to Amazon S3, so I'm going to pass on doing this implementation myself.

File: Library/Homebrew/archive.rb [hunk 1]
```diff
@@ -0,0 +1,185 @@
+# typed: false
+# frozen_string_literal: true
+
+require "digest/md5"
+require "utils/curl"
+
+# Archive API client.
```
Comment (MikeMcQuaid) [inline Library/Homebrew/archive.rb:7]:
```suggestion
# The Internet Archive API client.
```

File: Library/Homebrew/archive.rb [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    "#<Archive: item=#{@archive_item}>"
+  end
+
+  sig { params(item: T.nilable(String)).void }
+  def initialize(item: "homebrew")
+    @archive_item = item
+
+    raise UsageError, "Must set the Archive item!" unless @archive_item
+
+    ENV["HOMEBREW_FORCE_HOMEBREW_ON_LINUX"] = "1" if @archive_item == "homebrew" && !OS.mac?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/archive.rb:30]: Can you add a comment explaining this?

File: Library/Homebrew/archive.rb [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+  end
+
+  sig {
+    params(local_file:    String,
+           directory:     String,
+           remote_file:   String,
+           warn_on_error: T.nilable(T::Boolean)).void
+  }
+  def upload(local_file, directory:, remote_file:, warn_on_error: false)
+    unless File.exist? local_file
```
Comment (MikeMcQuaid) [inline Library/Homebrew/archive.rb:55]: Make `local_file` a `Pathname` and then you can use `local_file.exist?` here and `local_file.read` below.

File: Library/Homebrew/archive.rb [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+  include Context
+  include Utils::Curl
+
+  class Error < RuntimeError
+  end
+
+  sig { returns(String) }
+  def inspect
+    "#<Archive: item=#{@archive_item}>"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/archive.rb:21]: Would be good to get some tests for this class.

File: Library/Homebrew/archive.rb [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+    msg = "Bottle upload failed: #{result.stdout}"
+    raise msg unless warn_on_error
+
+    opoo msg
+  end
+
+  sig { params(url: String).returns(T::Boolean) }
+  def stable_mirrored?(url)
+    headers, = curl_output("--connect-timeout", "15", "--location", "--head", url)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/archive.rb:79]: Why 15? Would be good to have a comment or shared constant if this is copied from elsewhere.

File: Library/Homebrew/archive.rb [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+      msg = "#{local_file} for upload doesn't exist!"
+      raise Error, msg unless warn_on_error
+
+      # Warn and return early here since we know this upload is going to fail.
+      opoo msg
+      return
+    end
+
+    md5_base64 = Digest::MD5.base64digest(File.read(local_file))
+    url = "https://#{@archive_item}.s3.us.archive.org/#{directory}/#{remote_file}"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/archive.rb:65]: Would be good to make `s3.us.archive.org` a constant somewhere given it's used in a few places.

File: Library/Homebrew/archive.rb [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        ia delete #{@archive_item} #{directory}/#{filename}
+    EOS
+  end
+
+  sig {
+    params(bottles_hash:  T::Hash[String, T.untyped],
+           warn_on_error: T.nilable(T::Boolean)).void
+  }
+  def upload_bottles(bottles_hash, warn_on_error: false)
+    bottles_hash.each do |_formula_name, bottle_hash|
```
Comment (MikeMcQuaid) [inline Library/Homebrew/archive.rb:138]:
```suggestion
    bottles_hash.each_value do |bottle_hash|
```

File: Library/Homebrew/archive.rb [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+      directory = bottle_hash["bintray"]["repository"]
+      bottle_count = bottle_hash["bottle"]["tags"].length
+
+      bottle_hash["bottle"]["tags"].each do |_tag, tag_hash|
```
Comment (MikeMcQuaid) [inline Library/Homebrew/archive.rb:142]:
```suggestion
      bottle_hash["bottle"]["tags"].each_value do |tag_hash|
```

File: Library/Homebrew/dev-cmd/pr-pull.rb [hunk 9]
```diff
@@ -49,6 +49,8 @@ def pr_pull_args
              description: "Message to include when autosquashing revision bumps, deletions, and rebuilds."
       flag   "--artifact=",
              description: "Download artifacts with the specified name (default: `bottles`)."
+      flag   "--archive-item=",
+             description: "Upload to the specified Archive item (default: `homebrew`)."
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pr-pull.rb:53]:
```suggestion
             description: "Upload to the specified Internet Archive item (default: `homebrew`)."
```

File: Library/Homebrew/dev-cmd/pr-upload.rb [hunk 10]
```diff
@@ -27,6 +28,8 @@ def pr_upload_args
       switch "--warn-on-upload-failure",
              description: "Warn instead of raising an error if the bottle upload fails. "\
                           "Useful for repairing bottle uploads that previously failed."
+      flag   "--archive-item=",
+             description: "Upload to the specified Archive item (default: `homebrew`)."
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pr-upload.rb:32]:
```suggestion
             description: "Upload to the specified Internet Archive item (default: `homebrew`)."
```

File: Library/Homebrew/env_config.rb [hunk 11]
```diff
@@ -15,6 +15,10 @@ module EnvConfig
         description: "Linux only: Pass this value to a type name representing the compiler's `-march` option.",
         default:     "native",
       },
+      HOMEBREW_ARCHIVE_KEY:                   {
+        description: "Use this API key when accessing the Archive.org API (where bottles are stored). " \
```
Comment (MikeMcQuaid) [inline Library/Homebrew/env_config.rb:19]:
```suggestion
        description: "Use this API key when accessing the Internet Archive API (where bottles are stored). " \
```

File: Library/Homebrew/env_config.rb [hunk 12]
```diff
@@ -15,6 +15,10 @@ module EnvConfig
         description: "Linux only: Pass this value to a type name representing the compiler's `-march` option.",
         default:     "native",
       },
+      HOMEBREW_ARCHIVE_KEY:                   {
```
Comment (MikeMcQuaid) [inline Library/Homebrew/env_config.rb:18]:
```suggestion
      HOMEBREW_INTERNET_ARCHIVE_KEY:                   {
```

**Review by MikeMcQuaid — COMMENTED** [review]
Looking good so far, nice work!

File: Library/Homebrew/archive.rb (hunk 2 above)
Comment (sjackman) [reply Library/Homebrew/archive.rb:30]: This line is here on the foundational software engineering principles of cargo cult engineering and [Chesterton's Fence](https://en.wikipedia.org/wiki/Wikipedia:Chesterton's_fence), cribbed from https://github.com/Homebrew/brew/blob/3efab3efa7114afe713b2485032b740e6b55d6e4/Library/Homebrew/bintray.rb#L32
I'm not sure of its purpose, so I sadly cannot comment it. I'm happy deleting this line, and we can add it back if we find that we need it.

Comment (sjackman) [reply Library/Homebrew/archive.rb:30]: If I were to guess, it was a hack for uploading and verifying (more likely the verifying step) macOS bottles running on a Linux CI runner.

Comment (sjackman) [reply Library/Homebrew/archive.rb:30]: I've deleted this line. We can crib it from `bintray.rb` if needed.

File: Library/Homebrew/archive.rb (hunk 5 above)
Comment (sjackman) [reply Library/Homebrew/archive.rb:79]: 🤷‍♂️ Cribbed from `bintray.rb`

Comment (sjackman) [reply Library/Homebrew/archive.rb:79]: Deleted

File: Library/Homebrew/archive.rb (hunk 8 above)
Comment (sjackman) [reply Library/Homebrew/archive.rb:142]: Nice catch. Rubocop should have an audit for this case.

**Comment by sjackman** [comment]
BrewTestBot approved my PR, and it automerged. That wasn't exactly intentional, but I'm happy with this PR. Happy to follow up with another PR if tweaks are needed.

**Comment by MikeMcQuaid** [comment]
@sjackman pulling out some constants and adding any tests would be nice if possible!

**Pull request merged**